lxd/storage/backend_lxd.go

```diff
@@ -3731,19 +3735,135 @@ func (b *lxdBackend) ImportBucket(projectName string, poolVol *backupConfig.Conf
-	storageBucketName := project.StorageVolume(projectName, poolVol.Bucket.Name)
+	storageBucketName := project.StorageVolume(projectName, bucket.Name)
 	storageBucket := b.GetVolume(drivers.VolumeTypeBucket, drivers.ContentTypeFS, storageBucketName, bucketConfig)
 
 	err = b.driver.ValidateVolume(storageBucket, false)
 	if err != nil {
 		return nil, err
 	}
 
+	// Recover bucket keys.
+
+	// Start minio process.
+	minioProc, err := b.ActivateBucket(projectName, bucket.Name, op)
+	if err != nil {
+		return nil, err
+	}
+
+	// Initialize minio client object.
+	adminClient, err := minioProc.AdminClient()
+	if err != nil {
+		return nil, err
+	}
+
+	ctx, ctxCancel := context.WithTimeout(context.TODO(), time.Duration(time.Second*30))
+	defer ctxCancel()
+
+	// Export IAM data (response is ZIP file).
+	iamReader, err := adminClient.ExportIAM(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	defer iamReader.Close()
+
+	iamBytes, err := ioutil.ReadAll(iamReader)
+	if err != nil {
+		return nil, err
+	}
+
+	iamZipReader, err := zip.NewReader(bytes.NewReader(iamBytes), int64(len(iamBytes)))
+	if err != nil {
+		return nil, err
+	}
+
+	// We are interesed only in a json file that contains service accounts.
+	// Find that file and extract service accounts.
+	svcAccounts := map[string]madmin.Credentials{}
+	for _, file := range iamZipReader.File {
+		if file.Name != "iam-assets/svcaccts.json" {
+			continue
+		}
+
+		f, err := file.Open()
+		if err != nil {
+			return nil, err
+		}
+
+		defer f.Close()
+
+		fContent, err := ioutil.ReadAll(f)
+		if err != nil {
+			return nil, err
+		}
+
+		err = json.Unmarshal(fContent, &svcAccounts)
+		if err != nil {
+			return nil, err
+		}
+
+		break
+	}
+
+	adminPolicy, err := s3.BucketPolicy(bucket.Name, "admin")
+	if err != nil {
+		return nil, err
+	}
+
+	// Create bucket key DB entry for each service account.
+	for _, creds := range svcAccounts {
+		key := api.StorageBucketKeysPost{
+			Name: creds.AccessKey,
+			StorageBucketKeyPut: api.StorageBucketKeyPut{
+				Description: "Recovered bucket key",
+				Role:        "read-only",
+				AccessKey:   creds.AccessKey,
+				SecretKey:   creds.SecretKey,
+			},
+		}
+
+		svcAccountInfo, err := adminClient.InfoServiceAccount(ctx, creds.AccessKey)
+		if err != nil {
+			return nil, err
+		}
+
+		// If service account policy equals to the admin bucket policy, set bucket key role to admin.
+		if comparePolicy(string(adminPolicy), svcAccountInfo.Policy) {
+			key.Role = "admin"
+		}
+
+		keyID, err := b.state.DB.Cluster.CreateStoragePoolBucketKey(context.TODO(), bucketID, key)
+		if err != nil {
+			return nil, err
+		}
+
+		revert.Add(func() { _ = b.state.DB.Cluster.DeleteStoragePoolBucketKey(context.TODO(), bucketID, keyID) })
+	}
+
 	cleanup := revert.Clone().Fail
 	revert.Success()
 	return cleanup, err
 }
 
+// comparePolicy compares S3 policies and returns true if they are strictly equal.
+func comparePolicy(policyA, policyB string) bool {
```

Comment:
Can we move this to storage/utils.go and call it something clearer, like `minioPolicyCompare`?

Reply:
We dont use argument type compaction (not sure if thats the right terminology), so please can the function signature be:

```
comparePolicy(policyA string, policyB string) bool
```

Reply:
Yep sounds good! Does this only work for minio though or is it a general s3 thing?

Reply:
As far as I know, it is generic S3 policy.

Just as a note, comparing policies like this will work only for the roles that are predefined in the `lxd/storage/s3/policy.go`. Since any difference in the policy would result in invalid comparison. However, since the policy is managed exclusively by LXD I think this should not be a problem (at least from a perspective of bucket recovery).